src/utils/intrusive_ptr.h

```diff
@@ -33,6 +33,12 @@ class intrusive_ptr {
   intrusive_ptr(intrusive_ptr<T>&& other) : ptr_(other.ptr_) {
     other.ptr_ = nullptr;
   }
+  
+  template <class X, std::enable_if_t<std::is_convertible<X *, T *>::value, bool> = true>
+  intrusive_ptr(intrusive_ptr<X> other) : ptr_(other.get()) {
+    other.ptr_ = nullptr;
```

Comment:
This doesn't seem right. This is the behavior I would expect from a `move` constructor, where the new pointer "steals" the contents of the old pointer without incrementing the count.

This constructor's signature is closer to that of a copy constructor, and so it should increment the pointed-to reference counter.

In addition, if this is supposed to be a copy ctor, then the parameter type should be `const intrusive_ptr<X> &other`, so that the parameter points to the existing copy rather than silently making a copy, which would require altering the refcount.